libbpf-tools/readahead.c

```diff
@@ -119,28 +139,35 @@ int main(int argc, char **argv)
 		fprintf(stderr, "failed to open BPF object\n");
 		return 1;
 	}
-
-	/*
-	 * Starting from v5.10-rc1 (8238287), __do_page_cache_readahead has
-	 * renamed to do_page_cache_ra. So we specify the function dynamically.
-	 */
-	err = readahead__set_attach_target(obj->progs.do_page_cache_ra);
-	if (err)
-		goto cleanup;
-	err = readahead__set_attach_target(obj->progs.do_page_cache_ra_ret);
-	if (err)
-		goto cleanup;
-
 	err = readahead_bpf__load(obj);
 	if (err) {
 		fprintf(stderr, "failed to load BPF object\n");
 		goto cleanup;
 	}
 
-	err = readahead_bpf__attach(obj);
-	if (err) {
-		fprintf(stderr, "failed to attach BPF programs\n");
-		goto cleanup;
+	link = bpf_program__attach_trace(obj->progs.fexit_page_cache_alloc_ret);
+	if ((long)link == -524 /*ENOTSUPP*/) {
+		err = readahead__probe_target(obj->progs.kprobe_do_page_cache_ra, false);
+		if (err)
+			goto cleanup;
+		err = readahead__probe_target(obj->progs.kprobe_do_page_cache_ra, true);
+		if (err)
+			goto cleanup;
+		bpf_program__attach_kprobe(obj->progs.kretprobe_page_cache_alloc_ret,
```

Comment:
The return value of `bpf_program__attach_kprobe` should assign to `obj->links.xxx`, and we should do error check here.

Reply:
It will not work since bpf_program__attach_kprobe returns perf_event_open_probe's return value if the latter fails.

Reply:
> The return value of `bpf_program__attach_kprobe` should assign to `obj->links.xxx`, and we should do error check here.

It will not work since bpf_program__attach_kprobe returns perf_event_open_probe's return value if the latter fails.